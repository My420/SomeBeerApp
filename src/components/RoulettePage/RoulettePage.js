import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'immutable';
import { connect } from 'react-redux';
import styles from './RoulettePage.module.scss';
import IconLoadingSvg from '../IconLoadingSvg/IconLoadingSvg';
import ErrorMsg from '../ErrorMsg/ErrorMsg';
import LoadStatusSwitcher from '../LoadStatusSwitcher/LoadStatusSwitcher';
import getRouletteData from '../../ActionCreator/getRouletteData';
import RoulettePanel from '../RoulettePanel/RoulettePanel';

export class RoulettePage extends React.Component {
  componentDidMount() {
    const { getData, isLoaded } = this.props;
    if (!isLoaded) {
      getData();
    }
  }

  render() {
    // eslint-disable-next-line no-console
    console.log('render ============================ RoulettePage');

    const { isLoaded, isLoading, isError, errorMessage, data } = this.props;

    const LoadingComponent = () => {
      return (
        <div className={styles.loader}>
          <IconLoadingSvg color="#e69e63" />
        </div>
      );
    };

    const ErrorComponent = () => {
      return (
        <div className={styles.error}>
          <ErrorMsg errorMsg={errorMessage} />
        </div>
      );
    };

    const DataComponent = () => {
      return <RoulettePanel data={data} />;
    };

    return (
      <React.Fragment>
        <LoadStatusSwitcher
          isLoading={isLoading}
          isError={isError}
          isLoaded={isLoaded}
          LoadingComponent={LoadingComponent}
          ErrorComponent={ErrorComponent}
          DataComponent={DataComponent}
        />
      </React.Fragment>
    );
  }
}

const mapStateToProps = store => {
  return {
    isLoaded: store.roulette.isLoaded,
    isLoading: store.roulette.isLoading,
    isError: store.roulette.isError,
    errorMessage: store.roulette.errorMessage,
    data: store.roulette.data
  };
};

const mapDispatchToProps = dispatch => {
  return {
    getData: () => {
      dispatch(getRouletteData());
    }
  };
};

RoulettePage.propTypes = {
  isLoaded: PropTypes.bool.isRequired, // form connect
  isLoading: PropTypes.bool.isRequired, // form connect
  isError: PropTypes.bool.isRequired, // form connect
  errorMessage: PropTypes.string.isRequired, // form connect
  data: PropTypes.instanceOf(List).isRequired, // form connect
  getData: PropTypes.func.isRequired // form connect
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(RoulettePage);
