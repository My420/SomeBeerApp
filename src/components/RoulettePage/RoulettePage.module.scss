.loader {
  width: 100%;
  height: 80px;
}

.error {
  width: 100%;
  height: auto;

  margin-bottom: 20px;
}
