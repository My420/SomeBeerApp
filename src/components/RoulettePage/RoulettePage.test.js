import React from 'react';
import { shallow } from 'enzyme';
import { RoulettePage } from './RoulettePage';

describe('Test <RoulettePage /> component', () => {
  const props = {
    isLoading: false,
    isLoaded: false,
    isError: false,
    errorMessage: '',
    data: 'test',
    getData: () => {}
  };

  const wrapper = shallow(<RoulettePage {...props} />);
  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should render LoadStatusSwitcher component with correct props', () => {
    const { isLoaded, isLoading, isError } = props;
    expect(wrapper.find('LoadStatusSwitcher')).toHaveLength(1);
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoading')).toBe(
      isLoading
    );
    expect(wrapper.find('LoadStatusSwitcher').prop('isLoaded')).toBe(isLoaded);
    expect(wrapper.find('LoadStatusSwitcher').prop('isError')).toBe(isError);
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('LoadingComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('ErrorComponent')
    ).toBe('function');
    expect(
      typeof wrapper.find('LoadStatusSwitcher').prop('DataComponent')
    ).toBe('function');
  });

  describe('test getData function call', () => {
    test('call the `getBeerData` function in componentDidMount', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, getData: testFunc };
      shallow(<RoulettePage {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(1);
    });

    test('do not call the `getBeerData` function if isLoaded is true', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, isLoaded: true, getBeerData: testFunc };
      shallow(<RoulettePage {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(0);
    });

    test('call the `getData` function if isLoaded is false', () => {
      const testFunc = jest.fn();
      const newProps = { ...props, isError: true, getData: testFunc };
      shallow(<RoulettePage {...newProps} />);

      expect(testFunc.mock.calls.length).toBe(1);
    });
  });
});
