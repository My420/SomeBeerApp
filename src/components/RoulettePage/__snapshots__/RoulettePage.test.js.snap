// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <RoulettePage /> component should match snapshot 1`] = `
<Main
  className="main"
>
  <Container
    className="container"
  >
    <h2
      className="name"
    >
      <span
        className="shopName"
      >
        Some Beer Shop
      </span>
      <span
        className="pageName"
      >
        Roulette!
      </span>
    </h2>
    <p
      className="slogan"
    >
      We will decide which beer you get!
    </p>
    <LoadStatusSwitcher
      DataComponent={[Function]}
      ErrorComponent={[Function]}
      LoadingComponent={[Function]}
      isError={false}
      isLoaded={false}
      isLoading={false}
    />
  </Container>
</Main>
`;
